### ISSUE TITLE
Dell XPS 13 9333 touchpad doesn't work on kernel 3.18

### ISSUE THREAD

**Issue by sorokin** [NixOS/nixpkgs#9133]
After upgrade to the latest master my touchpad stopped working.

As I found out the problem is kernel 3.18, when I change kernel to 3.14 everything works ok.

I switch kernel versions by changing pkgs/top-level/all-packages.nix:

```
-  linuxPackages = linuxPackages_3_18;
+  linuxPackages = linuxPackages_3_14;
```

The problem is very similar to one reported in https://bugs.archlinux.org/task/40921 There they claim that bug is fixed in 3.16, but there is people who experience the same issue on 3.16 kernel: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1386077

I don't know how to debug this kind of issues. xinput simple doesn't list a touchpad:

```
[ivan@nixos:~]$ cat xinput.txt.3.14 
⎡ Virtual core pointer                          id=2    [master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer                id=4    [slave  pointer  (2)]
⎜   ↳ SYNAPTICS Synaptics Large Touch Screen    id=9    [slave  pointer  (2)]
⎜   ↳ SynPS/2 Synaptics TouchPad                id=12   [slave  pointer  (2)]
⎣ Virtual core keyboard                         id=3    [master keyboard (2)]
    ↳ Virtual core XTEST keyboard               id=5    [slave  keyboard (3)]
    ↳ Power Button                              id=6    [slave  keyboard (3)]
    ↳ Video Bus                                 id=7    [slave  keyboard (3)]
    ↳ Power Button                              id=8    [slave  keyboard (3)]
    ↳ Integrated_Webcam_HD                      id=10   [slave  keyboard (3)]
    ↳ AT Translated Set 2 keyboard              id=11   [slave  keyboard (3)]
    ↳ Dell WMI hotkeys                          id=13   [slave  keyboard (3)]

[ivan@nixos:~]$ cat xinput.txt.3.18
⎡ Virtual core pointer                          id=2    [master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer                id=4    [slave  pointer  (2)]
⎜   ↳ SYNAPTICS Synaptics Large Touch Screen    id=9    [slave  pointer  (2)]
⎣ Virtual core keyboard                         id=3    [master keyboard (2)]
    ↳ Virtual core XTEST keyboard               id=5    [slave  keyboard (3)]
    ↳ Power Button                              id=6    [slave  keyboard (3)]
    ↳ Video Bus                                 id=7    [slave  keyboard (3)]
    ↳ Power Button                              id=8    [slave  keyboard (3)]
    ↳ Integrated_Webcam_HD                      id=10   [slave  keyboard (3)]
    ↳ AT Translated Set 2 keyboard              id=11   [slave  keyboard (3)]
    ↳ Dell WMI hotkeys                          id=12   [slave  keyboard (3)]
```

I'm not sure if this report fit nixos bugtracker. Probably this is a kernel problem, but also it can be some module-blacklisting problem. So I'm reporting this issue against nixos as the immediate vendor. Probably this issue should be forwarded somewhere.

**Comment by sorokin** [comment]
I tested kernel 4.0 and 4.1 touchpad doesn't work there either.

**Comment by cpages** [comment]
What about dmesg? Does it show anything related to the touchpad?

**Comment by sorokin** [comment]
dmesg.3.14 http://pastebin.com/YeCR1rqk
dmesg.3.18 http://pastebin.com/xT1BmG9t

**Comment by sorokin** [comment]
/proc/bus/input/devices.3.14 http://pastebin.com/h2gxz5bE
/proc/bus/input/devices.3.18 http://pastebin.com/Z0QkE4zk

Kernel just silently ignore presence of a touchpad. :-(

**Comment by sorokin** [comment]
Here it a few my crazy ideas. Tell me if I'm going to do something completely wrong.

Probably I should try 3.15, 3.16, 3.17 kernels? Lookin at pkgs/os-specific/linux/kernel it seems it is relatively easy to add them. Testing them will get me closer to understanding where touchpad became broken.

Another idea is to bisect linux kernel to find out what commit introduced a problem. Can I use fetch url to take kernel sources from local git repository? I don't have an access to fast build machine and compiling on my laptop can take a while. I think I can reduce amount of code to be compiled by modifying common-config.nix.

**Comment by sorokin** [comment]
I tested kernel 3.15. Touchpad is broken here too.

**Comment by vcunat** [comment]
In your place I'd try searching cross-distro, e.g. have you tried [this](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1296743/comments/9)?

**Comment by sorokin** [comment]
> In your place I'd try searching cross-distro, e.g. have you tried this?

I tried writing in /etc/nixos/configuration.nix a line
```
  boot.blacklistedKernelModules = [ "i2c-hid" ];
```
It gave me a suggested line "blacklist i2c-hid" in /etc/modprobe.d/nixos.conf, but touchpad still didn't work. I didn't check if i2c-hid was present lsmod. When I build a new kernel, at the next reboot I will check lsmod (I have started bisecting).

**Comment by eskimor** [comment]
any news on this one? Does your keyboard work on xps 13? Mine stops working after a while after each boot. It works fine with the pre-installed ubuntu. Sigh.

**Comment by eskimor** [comment]
Hehe. It seems to work on 4.1.5 again! Just make sure you are not blacklisting psmouse. (Which was a suggested fix in the Arch wiki I think)

**Comment by cumber** [comment]
@sorokin, did you ever find a solution?

I've just installed NixOS 15.09 on a Dell 9333, and the touchpad doesn't work for me either (completely undetected). Blacklisting i2c-hid (or i2c_hid) doesn't make a difference, nor does switching to kernel 4.2.3. If I force the kernel version back to 3.14 then it does work.

**Comment by CMCDragonkai** [comment]
On NixOS live CD 15.09, touchpad doesn't work. However the touchpad is detected according to `xinput --list`. It shows `SynPS/2 Synaptics Touchpad`. A normal mouse connected via USB works though. The left and right buttons work, but you can't move the mouse.

**Comment by cumber** [comment]
I was able to get the touchpad working with a kernel newer than 3.14 in the end. I applied a BIOS patch from Dell (I had version A05 installed before, updated to A08). After that it worked with kernel 4.2, but still would not work with 3.18 (and still worked with 3.14). "Not works" here means it doesn't even show up in xinput (or the /dev tree, or device lists I found somewhere under /proc); all cases were the same.

All I was changing between the reboots was choosing one of the following in configuration.nix (or leaving it off entirely to get the default 3.18):

    linuxPackages = linuxPackages_3_14
    linuxPackages = linuxPackages_4_2
